splash/benchmark/benchmark.py

```diff
@@ -0,0 +1,169 @@
+PORT = 8806
+#: Combinations of width & height to test.
+WIDTH_HEIGHT = [(None, None), (500, None), (None, 500), (500, 500)]
+#: Splash log filename.
+SPLASH_LOG = 'splash.log'
+#: This script is used to collect maxrss & cpu time from splash process.
+GET_PERF_STATS_SCRIPT = """
+function main(splash)
+  return splash:get_perf_stats()
+end
+"""
+
+
+parser = ArgumentParser(description=__doc__,
+                        formatter_class=ArgumentDefaultsHelpFormatter)
+parser.add_argument('--seed', type=int, default=1234, help='PRNG seed number')
+parser.add_argument('--thread-count', type=int, default=1,
+                    help='Request thread count')
+parser.add_argument('--request-count', type=int, default=10,
+                    help='Benchmark request count')
+parser.add_argument('--sites-dir', type=str, default='sites',
+                    help='Directory with downloaded sites')
+
+
+def generate_requests(splash, args):
+    log = logging.getLogger('generate_requests')
+    log.info("Using pRNG seed: %s", args.seed)
+
+    # Static pages (relative to sites_dir) to be used in the benchmark.
+    pages = [re.sub('^%s/' % args.sites_dir, '', v)
+             for v in glob(os.path.join(args.sites_dir, 'localhost_8806',
+                                        '*.html'))]
+    for p in pages:
+        log.info("Using page for benchmark: %s", p)
+
+    rng = random.Random(args.seed)
+    for i in xrange(args.request_count):
+        page = rng.choice(pages)
+        width, height = rng.choice(WIDTH_HEIGHT)
+        req_factory = rng.choice(REQ_FACTORIES)
+        url = 'http://localhost:%d/%s' % (PORT, page)
+        params = {'url': url, 'render_all': 1, 'wait': 0.1,
+                  'width': width, 'height': height}
+        log.debug("Req factory: %s, params: %s", req_factory, params)
+        yield (i + 1, args.request_count, req_factory(splash, params))
+
+
+def parallel_map(func, iterable, thread_count):
+    if thread_count == 1:
+        return map(func, iterable)
+    else:
+        pool = ThreadPool(thread_count)
+        return pool.map(func, iterable)
+
+
+def invoke_request(invoke_args):
+    log = logging.getLogger('bench_worker')
+    req_no, total_reqs, kwargs = invoke_args
+    log.info("Initiating request %d/%d: %s", req_no, total_reqs, kwargs)
+    stime = time()
+    response = requests.get(**kwargs)
+    etime = time()
+    if response.status_code != 200:
+        log.error("Non-OK response:\n%s", response.text)
+    return {'start_time': stime,
+            'end_time': etime,
+            'duration': etime - stime,
+            'endpoint': kwargs['url'],
+            'status': response.status_code,
+            'site': kwargs['params']['url'],
+            'width': kwargs['params']['width'],
+            'height': kwargs['params']['height']}
+
+
+def main():
+    log = logging.getLogger("benchmark")
+    args = parser.parse_args()
+    logging.basicConfig(level=logging.DEBUG)
+
+    splash = SplashServer(
```

Comment:
It is better to allow to set Splash server address from the command line. This way it'll be easier to compare performance of different versions of Splash, or performance of dockerized Splash (maybe several dockerized Splash'es which use different base operating systems and/or qt/... versions).

We should really allow to run the same benchmark against different Splash versions. If the benchmark can only be executed against current Splash version we can't just checkout to older Splash version and run a benchmark again - checking out old Splash could also change the benchmark version, so we'll run a different benchmark.